### PR TITLE
Use FINAL declarations for immutable variables

### LINKED ISSUE #285: FINAL new ABAP keyword - guidance
Hi colleagues,
recently the new FINAL keyword has been introduced. Do you plan to document a common guidance for it? 
In my opinion this was something that I missed in ABAP for long time. 

Thank you,
Br. Petr

### PR DESCRIPTION
**Pull request by ConjuringCoffee**
Fixes #285.

I've added the new rule and changed all examples in which it was obvious that a variable is never reassigned.

### CODE REVIEW

**Review by cribuc — APPROVED** [review]

**Review by N2oB6n-SAP — CHANGES_REQUESTED** [review]
I have suggested a change to address the subtlety of `FINAL` that it does not prevent changes in general. It merely prevents changes in unexpected places.

File: clean-abap/CleanABAP.md [hunk 1]
```diff
@@ -962,6 +963,13 @@ ENDIF.
 
 > Read more in _Chapter 5: Formatting: Vertical Distance: Variable Declarations_ of [Robert C. Martin's _Clean Code_].
 
+### Use FINAL declarations for immutable variables
+
+> [Clean ABAP](#clean-abap) > [Content](#content) > [Variables](#variables) > [This section](#use-final-declarations-for-immutable-variables)
+
+If a variable is never reassigned, use `FINAL` to declare it. 
+The reader can then be sure that the content of the variable will never change.
```
[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md:971]: Think mutation of a final variable in subsequent loop passes, also see [hints in the keyword documentation](https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/index.htm?file=abenfinal_inline.htm).

**Comment by ConjuringCoffee** [comment]
Now here's a problem: `FINAL` doesn't just prevent changes in unexpected locations in the source code, it also affects the debugger. You can't change the variable in the debugger anymore.

I find this to be a major downside of the new keyword. Does anyone know if there's a workaround or a setting for the debugger? If not, then I'm personally quite conflicted about the new rule. I've converted the pull request to a draft until further clarification took place.

**Comment by ChristophStoeckSAP** [comment]
One comment form the ABAP Debugger perspective:
There is no way to change FINAL variables in the ABAP debugger. This is the same as with constants.

**Comment by bjoern-jueliger-sap** [comment]
The issue with `final` is a bit more subtle than "use it if a variable is never reassigned", and the fact that final variables cannot be changed in the debugger (and that this is _surprising_ to at least some users of the feature) hints at the underlying mismatch in concepts of immutability:

While it may be tempting to think about `final` as a hint to the compiler to emit syntax errors when you try to change that variable again, `final` variables in fact have different _runtime_ properties than normal variables: The notion of immutability is a _runtime_ property in ABAP, not a compile-time property! 

The following code is syntax-error free, but produces a short dump at runtime:

```abap
final(var) = 1.
assign var to field-symbol(<var>).
<var> = 11.
```

Changing the data declaration to use `data` instead of `final` avoids the short dump.

This is probably not what you expect if you are used to thinking about immutability as a compile-time property of _bindings_ (like const variables in Javascript or non-mut variables in Rust), even though the [documentation](https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/index.htm?file=abenfinal_inline.htm) of `final` says it: "In all other positions, any write access leads either to a syntax error or the uncatchable exception MOVE_TO_LIT_NOTALLOWED_NODATA." It's important to realize that "write access" refers to the _data object_ created by the `final` declaration, not to write access to the variable name (the binding).

In the example above it is easy to understand why the dump happens once you realize `final` sets a run-time property on the data object that underlies `var`, but if the dump involves a call stack and references, this can get difficult to analyze. For instance, it is not a syntax error to use a reference to a `final` variable as the actual parameter of another method:

```abap
final(var) = 1.
not_my_method( ref #( var ) ).
```

If `not_my_method` then at any point attempts to change the content of the reference, we get a short dump - even though `var` is never "reassigned" in the part of the program that is under our control.

So while I think a recommendation in the spirit of "Use `final` whenever appropriate" should be made, I think any such recommendation needs to have some details about what "appropriate" means: Especially when references are involved, the programmer needs to think carefully about whether or not the write-protection that `final` generates at runtime is intended or not.

**Comment by ConjuringCoffee** [comment]
Great comment. I am closing this pull request and pointed the issue to it.

**Pull request closed without merging**